pkg/kubelet/container/runtime.go

```diff
@@ -335,6 +348,8 @@ type Status struct {
 	// Message written by the container before exiting (stored in
 	// TerminationMessagePath).
 	Message string
+	// CPU and memory resources for this container
```

Comment:
@mrunalp @haircommander do you see any concerns if we poll the runtime for applied resources when calling getContainerStatus on every invocation?  i assume this is a free check at runtime side by just looking at cached container config?

Reply:
I'm assuming this is the configured value at the runtime level, and not the current value that would be read from the cgroups. If that's the case, then yes, we already cache this and it is trivial to return

Reply:
The E2E tests @wangchen615 and I developed currently test against the actual cgroup values converging to the desired values expressed by UpdateContainerResources CRI. Currently, check for convergence of actual values in ContainerStatus response is disabled as we need runtime support (it is beta blocker). If the runtime reports cached configured values as opposed to actuated cgroup values, we may fail the test.
Is it non-trivial to return that values that's actuated and in-effect on the container?

Reply:
@haircommander @mrunalp can you reply @vinaykul comment?

This would imply we need a cached value that captures the actually allocated value on the host, not just the desired value.  The use case where the two values are most likely to differ are for incompressible resources like memory.  cgroups v1 will not actually record an update to memcg values if the usage is greater than the supplied value.

This is core to the function of this use case so we know usage has actually converged with desired state.

We need to discuss this in SIG Node as I fear there is an oversight here.

Reply:
by "actual cgroup value" I mean the current usage by the container. AFAIK, if we call a `runc update` the values will either be updated, or the command will fail. thus, the limit we cache (that which we used to do the update) should reflect the state of the cgroup *limit*, but not necessarily the amount of that resource is currently used. If you're just checking the value of the cgroup for the limit, then it seems to me that it would work as expected using the cached value in the CRI

Reply:
hm https://github.com/kubernetes/kubernetes/pull/102884#discussion_r896976528 implies the opposite: that we should be returning the usage value from the cgroup. we still can do that, but it could be more expensive (or take longer to update, as we're currently updating the memory/cpu usage on a periodic basis)

Reply:
@haircommander @mrunalp @derekwaynecarr I'm not looking for the current usage (the value returned in ContainerStats to be clear). I'm interested in the value that is being enforced.

Consider the following example: If I have a newly created & running single-container pod with memory.request=memory.limit=1000 bytes and the container application has malloc'd 245 bytes, I'm not looking for 245 (value returned in MemoryUsage.usage_bytes if I understood correctly) at any point. I'm expecting the response to give me 1000.

Now, if I updated the memory limit to 500 bytes (UpdateContainerRequest should succeed), the next ContainerStatus response should return 500 bytes if my attempt to malloc(255) succeeds but malloc(256) doesn't end well - (OOM kill IIRC). If the allocation of 245+256=501 bytes does not result in an OOM kill (eventually), then presumably memory.limit_in_bytes=1000 byte is still in force, and the response should be 1000.

Response to UpdateContainerResources can still be 'success' if you want to cache and apply later, but the ask is you respond to ContainerStatus with 500 only after 'runc update' has succeeded and 500 byte limit is in force. Is this possible?

Reply:
@vinaykul just closing out this thread, agree that we clarified.

Reply:
SG. Resolving conversation.